Fix complex include filters

MRichards99 opened the pull request:
This PR will close #234 

## Description
As per the issue, a complex include filter input didn't work in the ICAT backend, but did work in the DB backend. This is a small change to fix that, with a couple of tests added to demonstrate this.

## Testing Instructions
Check the request in the issue (and ones similar to it) now work and don't return errors.

- [ ] Review code
- [ ] Check GitHub Actions build
- [ ] Review changes to test coverage
- [ ] {more steps here}

## Agile Board Tracking
Connect to #234

sam-glendenning reviewed (commented):
So I tried that specific query out and I seemed to get different results. Both give a lot of output but I put it into a JSON diff tool and got the below (`db` backend on the left, `python_icat` on the right):

![diff](https://user-images.githubusercontent.com/71134574/119175745-21b4f200-ba62-11eb-99ab-7c42a6896c25.PNG)

There's many more differences beyond this image, this is just a snapshot. Happy to demonstrate this over Zoom but I'd like to see what anyone else got in case I'm missing something

MRichards99 commented:
So the foreign keys on the left (e.g. `facilityID`) will hopefully be removed in #224. `pid` inside `instrument` is there because the DB backend is based on a version of ICAT Server older than 4.10.0 (which is around then that attribute was introduced and the same thing happens with `user.affiliation`. I'm not sure the exact version the DB backend is based on, but 4.8.0 or 4.9.0 would be my guess. Should've put those details in the PR but I was trying to get it done by the end of Friday afternoon :) I think those are the only differences, unless I've missed some others?

I've just pushed a commit which fixes a bug I introduced on this branch - just a small mistake with `join()` - I've added a test case to monitor that too.

sam-glendenning commented:
> So the foreign keys on the left (e.g. `facilityID`) will hopefully be removed in #224. `pid` inside `instrument` is there because the DB backend is based on a version of ICAT Server older than 4.10.0 (which is around then that attribute was introduced and the same thing happens with `user.affiliation`. I'm not sure the exact version the DB backend is based on, but 4.8.0 or 4.9.0 would be my guess. Should've put those details in the PR but I was trying to get it done by the end of Friday afternoon :) I think those are the only differences, unless I've missed some others?
> 
> I've just pushed a commit which fixes a bug I introduced on this branch - just a small mistake with `join()` - I've added a test case to monitor that too.

@MRichards99 As long as the differences I've mentioned can be accounted for, I'm happy to approve it. Although your codecov isn't happy...

MRichards99 commented:
Yeah I can't see anything else that concerns me. Overall codecov diff is -0.01% now so pretty marginal
